types/lib/utils/flattened-nodes-observer.d.ts

```diff
@@ -0,0 +1,90 @@
+/**
+ * DO NOT EDIT
+ *
+ * This file was automatically generated by
+ *   https://github.com/Polymer/gen-typescript-declarations
+ *
+ * To modify these typings, edit the source file(s):
+ *   lib/utils/flattened-nodes-observer.html
+ */
+
+/// <reference path="boot.d.ts" />
+/// <reference path="array-splice.d.ts" />
+/// <reference path="async.d.ts" />
+
+declare namespace Polymer {
+
+  /**
+   * Class that listens for changes (additions or removals) to
+   * "flattened nodes" on a given `node`. The list of flattened nodes consists
+   * of a node's children and, for any children that are `<slot>` elements,
+   * the expanded flattened list of `assignedNodes`.
+   * For example, if the observed node has children `<a></a><slot></slot><b></b>`
+   * and the `<slot>` has one `<div>` assigned to it, then the flattened
+   * nodes list is `<a></a><div></div><b></b>`. If the `<slot>` has other
+   * `<slot>` elements assigned to it, these are flattened as well.
+   *
+   * The provided `callback` is called whenever any change to this list
+   * of flattened nodes occurs, where an addition or removal of a node is
+   * considered a change. The `callback` is called with one argument, an object
+   * containing an array of any `addedNodes` and `removedNodes`.
+   *
+   * Note: the callback is called asynchronous to any changes
+   * at a microtask checkpoint. This is because observation is performed using
+   * `MutationObserver` and the `<slot>` element's `slotchange` event which
+   * are asynchronous.
+   *
+   * An example:
+   * ```js
+   * class TestSelfObserve extends Polymer.Element {
+   *   static get is() { return 'test-self-observe';}
+   *   connectedCallback() {
+   *     super.connectedCallback();
+   *     this._observer = new Polymer.FlattenedNodesObserver(this, (info) => {
+   *       this.info = info;
+   *     });
+   *   }
+   *   disconnectedCallback() {
+   *     super.disconnectedCallback();
+   *     this._observer.disconnect();
+   *   }
+   * }
+   * customElements.define(TestSelfObserve.is, TestSelfObserve);
+   * ```
+   */
+  class FlattenedNodesObserver {
+    _shadyChildrenObserver: MutationObserver|null;
+    _nativeChildrenObserver: MutationObserver|null;
+    _boundSchedule: () => any;
```

Comment:
Hm are these actually supposed to be part of the public API @kevinpschaaf @sorvell ? I do not see them mentioned on https://www.polymer-project.org/2.0/docs/api/classes/Polymer.FlattenedNodesObserver

Reply:
Yeah these are probably meant to be private or protected, as denoted by the `_` prefix

Reply:
Basically, these are intended to be "private".

On element classes/mixins, we have followed a convention of `_` single underscore is "protected" (intended for subclassers not instance users, but we will maintain the API per semver) and `__` as "private" (implementation detail, only the author should use, can change at any time, no semver guarantee).

We were a little sloppier on non-element classes, since we basically have way less expectation that anyone will subclass them, but that's not really defensible.  At a minimum we should `@private` annotate these, if not follow the same `__` convention.

Reply:
`@private` seems like the way to go, IMO

Reply:
Sounds good.

I just pulled all of the JSDoc changes, including these new `@private` annotations, into a separate PR at https://github.com/Polymer/polymer/pull/4997/files.